### PR TITLE
 video emergency hotfix

### PR DESCRIPTION
**Pull request by zoff99**
to get at least 65k bytes (and not less) out of those crippled packets, apply this emergency fix to 0.1.10

---
This change is [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/622)

### CODE REVIEW

**Comment by zoff99** [comment]
please squash ("Allow edits from maintainers" is on)

**Comment by nurupo** [comment]
Removed the "emergency" part and added text that it breaks protocol. "Emergency" implies that it should be merged asap, but we can't really do that without any planning as this would break the protocol. We might even have to wait out a release or two before merging this change, depending on what we decide to do with it.

**Comment by nurupo** [comment]
Btw, I know I have already commented on it months ago, but your git workflow is horrible :\ All these merge commits of PRs with c-toxcore changes, when you could have just rebased your repo on TokTok/c-toxcore's master without any extra merge commits. Not sure if this is manageable, haven't dealt with such PRs before, someone might have to manually rewrite your fixes and create a new clean PR with them.

**Comment by SkyzohKey** [comment]
@nurupo Just squash it and it should be fine, nope? As only one file changed

**Comment by nurupo** [comment]
Hm, in the issue #620 you changed struct members, but in here you just bound some values? What is the purpose of bounding them?

**Comment by nurupo** [comment]
Also, why number 65533? Why not 65535, for example? I'm not familiar with toxav code, so I might be missing something obvious.

**Comment by zoff99** [comment]
@nurupo 
please first understand, read. and then change stuff.

this is the emergency hotfix, that does NOT break protocol!
65535 will not work here. its at least 1 less. but to be sure i made it 2 less.

**Comment by nurupo** [comment]
Yeah, sorry, I assumed it was the same fix you linked in the issue without looking at the code.

**Comment by zoff99** [comment]
this could be released (after testing) quickly as 0.10.1 hotfix. so that we at least salvage what we have now out there

**Comment by mannol** [comment]
I recommend trying [this](https://github.com/TokTok/c-toxcore/issues/620#issuecomment-346902071) first, before making such a big change.

**Comment by nurupo** [comment]
Reviewed 1 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[toxav/toxav.c, line 721 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/622#-Kzk4pPhKPNvjmeBolbl:-Kzk4pPiUBfiqztDyM2i:b-i45sqb) ([raw file](https://github.com/toktok/c-toxcore/blob/0bc2c161de96aabdc162e74397d7910f09fdd5b9/toxav/toxav.c#L721)):*
> ```C
>         }
> 
> 		uint32_t frame_size_safe = vrc + sizeof(sampling_rate);
> ```

Let's do bound checking to make sure no overflow can occur here.

- `vrc` is of type `int`, meaning it's at least 16 bits.
- `vrc` can't be negative, we checked this above.
- `vrc` is signed, so it has only half of its positive range.
- `sizeof(sampling_rate)` is 4.

There might be an issue if `int` is 64-bit or greater. A signed 64-bit integer can go as far as `(2^63)-1`, while `uint32_t` can hold at most `(2^32)-1`, meaning that `frame_size_safe` might overflow due to possible type narrowing.

The safe version of this code would be to do

```c
#include <stdint.h>

// make sure vrc + sizeof(sampling_rate) <= UINT16_MAX - 2
uint16_t frame_size_safe;
if (vrc <= UINT16_MAX - 2 - sizeof(sampling_rate)) {
    frame_size_safe = vrc;
} else {
    frame_size_safe = UINT16_MAX - 2;
}
```

or with casts, promoting non-negative int `vrc` to unsigned int, effectively doubling its range which would prevent an overflow due to an addition, and `UINT16_MAX` to unsigned int

```c
#include <stdint.h>

uint16_t frame_size_safe;
if ((unsigned)vrc + sizeof(sampling_rate) <= (unsigned)UINT16_MAX - 2) {
    frame_size_safe = vrc + sizeof(sampling_rate);
} else {
    frame_size_safe =  UINT16_MAX - 2;
}
```

---

*[toxav/toxav.c, line 823 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/622#-KzkFrms_06AaVvamDZ3:-KzkFrms_06AaVvamDZ4:bgmylz1) ([raw file](https://github.com/toktok/c-toxcore/blob/0bc2c161de96aabdc162e74397d7910f09fdd5b9/toxav/toxav.c#L823)):*
> ```C
>         while ((pkt = vpx_codec_get_cx_data(call->video.second->encoder, &iter))) {
>             if (pkt->kind == VPX_CODEC_CX_FRAME_PKT) {
>                 uint32_t frame_size_safe = pkt->data.frame.sz;
> ```

Let's do the check here too.

- `pkt->data.frame.sz` is of type `size_t`, so it's not a fixed type and it can vary from platform to platform.
- It's unsigned

There might occur an overflow if `size_t` is greater than 32-bits, so a safer way would be to do

```c
uint16_t frame_size_safe;
if (pkt->data.frame.sz <= UINT16_MAX - 2) {
    frame_size_safe = data.frame.sz;
} else {
    frame_size_safe = UINT16_MAX - 2;
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/622)*

**Comment by nurupo** [comment]
You could argue that Opus would never return a value bigger than `INT32_MAX` or that `pkt->data.frame.sz` is always at most `UINT32_MAX`, but I like to stay on the safe side and avoid any potential overflows, and going purely off type ranges those could occur, no matter how unlikely it is.

**Comment by nurupo** [comment]
Huh, the network test keeps failing and failing on Travis, this is like the fourth time Travis runs. I wonder if it has something to do with the recent network code changes, this failure seems to be way too consistent than your regular random test failures on Travis.

File: toxav/toxav.c [hunk 1]
```diff
@@ -814,13 +824,22 @@ bool toxav_video_send_frame(ToxAV *av, uint32_t friend_number, uint16_t width, u
...
+                    frame_size_safe = pkt->data.frame.sz;
+                } else {
+                    frame_size_safe = UINT16_MAX - 2;
+                }
 
-                pthread_mutex_unlock(call->mutex_video);
-                LOGGER_WARNING(av->m->log, "Could not send video frame: %s\n", strerror(errno));
-                rc = TOXAV_ERR_SEND_FRAME_RTP_FAILED;
-                goto END;
+                if (rtp_send_data(call->video.first, (const uint8_t *)pkt->data.frame.buf, pkt->data.frame.sz, av->m->log) < 0) {
```
Comment (zoff99) [inline toxav/toxav.c]: you are not using ***frame_size_safe*** here

Comment (nurupo) [reply toxav/toxav.c]: Good catch, fixed.

**Comment by nurupo** [comment]
I wonder if mannol wants to do some other patch for this. Splitting a frame into pieces sounds better than just truncating the frame and throwing away the rest.

**Comment by zoff99** [comment]
just for documentation, this is not the PR i wrote. it has been rewritten.

**Comment by nurupo** [comment]
That's right, I have rewritten git history to remove lots of merge commits, removing potential integer overflows, addeding some comments and writing a commit message on the way. You can still find the original code in Reviewable, it's pretty similar and it wouldn't happen without zoff finding the issue and making the PR fixing it, so it only makes sense to keep zoff's commit authorship.

**Comment by nurupo** [comment]
Reviewed 1 of 1 files at r4, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[toxav/toxav.c, line 721 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/622#-Kzk4pPhKPNvjmeBolbl:-KzudEmzAvkE0IooJQEs:b-u2j45q) ([raw file](https://github.com/toktok/c-toxcore/blob/0bc2c161de96aabdc162e74397d7910f09fdd5b9/toxav/toxav.c#L721)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

Let's do bound checking to make sure no overflow can occur here.

- `vrc` is of type `int`, meaning it's at least 16 bits.
- `vrc` can't be negative, we checked this above.
- `vrc` is signed, so it has only half of its positive range.
- `sizeof(sampling_rate)` is 4.

There might be an issue if `int` is 64-bit or greater. A signed 64-bit integer can go as far as `(2^63)-1`, while `uint32_t` can hold at most `(2^32)-1`, meaning that `frame_size_safe` might overflow due to possible type narrowing.

The safe version of this code would be to do

```c
#include <stdint.h>

// make sure vrc + sizeof(sampling_rate) <= UINT16_MAX - 2
uint16_t frame_size_safe;
if (vrc <= UINT16_MAX - 2 - sizeof(sampling_rate)) {
    frame_size_safe = vrc;
} else {
    frame_size_safe = UINT16_MAX - 2;
}
```

or with casts, promoting non-negative int `vrc` to unsigned int, effectively doubling its range which would prevent an overflow due to an addition, and `UINT16_MAX` to unsigned int

```c
#include <stdint.h>

uint16_t frame_size_safe;
if ((unsigned)vrc + sizeof(sampling_rate) <= (unsigned)UINT16_MAX - 2) {
    frame_size_safe = vrc + sizeof(sampling_rate);
} else {
    frame_size_safe =  UINT16_MAX - 2;
}
```

</blockquote></details>

That first code fragment should have said `frame_size_safe = vrc + sizeof(sampling_rate);` instead of just `frame_size_safe = vrc;`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/622)*

**Comment by robinlinden** [comment]
Closing this as #623 fixes the same issue by sending the payload in several packets instead of truncating it.

**Pull request closed without merging**